### ISSUE TITLE
Neo4j returns wrong results

### ISSUE THREAD

**Issue by Aurora-yzj** [neo4j/neo4j#13359]
Hi! I discovered a bug that Neo4j returns wrong results when dealing with my queries. The result of `a4` column should be `true` instead of `WEJ6Mifdo`. The detailed queries is listed below. 
**Neo4j Version:** 5.14.0 Community
**Operating System:** Ubuntu 22.04
**API:** Docker

### Steps to reproduce
1. Pull the image: `docker pull neo4j`
2. Create a directory on the host that will be a mount for the data: `mkdir $HOME/neo4j-data`
3. Start a new container that mounts to this directory: `sudo docker run -d --name neo4j-test -p 17474:7474 -p 17687:7687  -v $HOME/neo4j-data:/data neo4j`
4. Go to the web interface at "http://127.0.0.1:17474", login and execute the following series of queries to build the database:
```
CREATE INDEX i17 IF NOT EXISTS FOR (n:L5) ON (n.k33);
CREATE INDEX i18 IF NOT EXISTS FOR (n:L5) ON (n.k34);
CREATE (n0 :L11:L5{k71 : true, k70 : false, k73 : false, k72 : false, k31 : false, k74 : 2089636047, k33 : -1809381724, k32 : 1340048950, k68 : "CRebh", k35 : 1510320085, k34 : -745811966, k37 : "p8PhH", k69 : "CCWFvyZ8g", k36 : "NYPEX", id : 1});
CREATE (n0 {id : 2});
CREATE (n0 :L11:L5:L4{k71 : false, k70 : true, k73 : false, k72 : false, k31 : true, k74 : -83809908, k30 : -1504108816, k33 : 1821686742, k32 : -633333682, k68 : "AIHGJ69K", k35 : 1090051030, k34 : -842771084, k37 : "B3bCxKNd", k26 : true, k69 : "26PreC", k25 : -1130071991, k28 : "GylvsEFJf", k27 : false, k29 : false, id : 3});
CREATE (n0 :L11{k71 : true, k70 : true, k73 : false, k72 : true, k74 : 844948913, id : 5, k68 : "ahbom7v", k69 : "aYCoF"});
CREATE (n0 :L11:L5{k71 : false, k70 : false, k73 : false, k72 : false, k74 : -1881955020, id : 6, k32 : -1397765415, k68 : "tYrQ4", k35 : 850464827, k34 : 2085380180, k37 : "mXhdTFb", k69 : "XU4GuvVWy"});
CREATE (n0 :L11:L5{k71 : true, k70 : false, k31 : true, k74 : -458374297, k33 : -955212432, k32 : 1676360630, k68 : "i8bHwsE", k35 : 65233702, k34 : -382115658, k37 : "byRTfvOg", k69 : "MhLHE60r", k36 : "EhPZDcW", id : 8});
CREATE (n0 :L11{k71 : false, k70 : true, k73 : false, k72 : false, k74 : 985326677, id : 9, k68 : "nTuHmCnY"});
CREATE (n0 {id : 10});
CREATE (n0 :L11{k70 : true, k73 : false, k72 : true, k74 : -2095608746, id : 11, k68 : "WEJ6Mifdo", k69 : "us9k1KmB"});
CREATE (n0 :L11:L5{k71 : false, k70 : true, k73 : false, k72 : false, k31 : false, k74 : 49439310, k33 : -1469710255, k32 : -1670796292, k68 : "YhXpW6", k35 : -300591470, k34 : -1236901384, k37 : "gPucn", k69 : "PSH4wmmK9", k36 : "GdAEX01MF", id : 12});
CREATE (n0 :L11:L5{k71 : true, k70 : false, k72 : false, k31 : false, k74 : 1971665443, k33 : -653772835, k32 : -800494660, k68 : "hDE5w", k35 : 1992595986, k34 : -1240040982, k37 : "j3cr38", k69 : "TxPWXjs", k36 : "Sa1XTK", id : 13});
CREATE (n0 :L11:L5{k71 : true, k70 : true, k73 : true, k72 : true, k31 : true, k74 : 712547352, k33 : 1749553879, k32 : 458591600, k35 : 1624061302, k34 : 1343991316, k37 : "Xp3Cd7", k69 : "WhW8kIwb", k36 : "ZqgpH", id : 14});
MATCH (n0 {id : 13}), (n1 {id : 6}) MERGE(n0)-[r :T3{k187 : true, k186 : -1852139411, k189 : "h3sg7KXi", k188 : false, id : 16, k190 : "f7Xhkmkb", k191 : false}]->(n1);
MATCH (n0 {id : 1}), (n1 {id : 3}) MERGE(n0)-[r :T3{k187 : false, k186 : 1795019972, k189 : "4n2TS", k188 : false, id : 17, k190 : "v6z5e", k191 : true}]->(n1);
MATCH (n0 {id : 2}), (n1 {id : 6}) MERGE(n0)-[r :T3{k187 : false, k186 : -984262157, k189 : "CmHEXZLy", k188 : false, id : 18, k190 : "iVe4IiFT", k191 : false}]->(n1);
MATCH (n0 {id : 12}), (n1 {id : 9}) MERGE(n0)-[r :T3{k187 : false, k186 : 118278019, k189 : "hsXzLgQx", k188 : false, id : 20, k190 : "VhUBAS"}]->(n1);
MATCH (n0 {id : 6}), (n1 {id : 11}) MERGE(n0)-[r :T3{k187 : true, k186 : -565895201, k189 : "8togT", k188 : false, id : 21, k190 : "N09mm", k191 : true}]->(n1);
MATCH (n0 {id : 10}), (n1 {id : 5}) MERGE(n0)-[r :T3{k187 : true, k186 : 706647661, k189 : "xyBzm", id : 23, k190 : "Fu6DCUu", k191 : true}]->(n1);
```
5. Run the query (for the ease of analysis, I have expanded several predicates, and this will not affect the reproduce of the bug):
```
MATCH (n0 :L11)<-[r0 :T3]-(n1) WHERE (((NOT (NOT (n0.k70))) OR (NOT ((n0.k74) > (n0.k74)))) AND ((((n0.k68) = 'ahbom7v') AND ((r0.k190) = 'Fu6DCUu')) AND ((n1.id) = 10))) 
UNWIND [(r0.k186), 557243387] AS a0 
MATCH (n2 :L11 :L5)-[r1 :T3]->(n3 :L11), (n4 :L11)<-[r2 :T3]-(n5 :L11 :L5)<-[r3 :T3]-(n6), (n7 :L11 :L5)-[r4 :T3]->(n8 :L11 :L5 :L4) WHERE (
    (
        (
            (((((n7.k36) STARTS WITH 'NYP')) AND (NOT ((n4.k68) = (n4.k68)))) AND (NOT (627830378 <= (n5.k35)))) 
            OR 
            (((r0.k191)))
        )
    ) AND ((((((((((((n2.k36) = 'GdAEX01MF') AND ((r1.k186) = 118278019)) AND ((n3.k74) = 985326677)) AND ((n4.k74) = -2095608746)) AND ((r2.id) = 21)) AND ((n5.k35) = 850464827)) AND ((r3.k189) = 'CmHEXZLy')) AND ((n6.id) = 2)) AND ((n7.k33) = -1809381724)) AND ((r4.id) = 17)) AND ((n8.k74) = -83809908))
    ) 
RETURN (r4.k190) AS a3, (r4.k191) AS a4
```

### Expected behavior
Please note that the condition for selecting r4 includes `(r4.id) = 17`:
```
...
WHERE (
   ... AND 
(((((((((( ... AND ((r4.id) = 17)) AND ...
    ) 
RETURN (r4.k190) AS a3, (r4.k191) AS a4
```
 and the only matching relation would be:
`MATCH (n0 {id : 1}), (n1 {id : 3}) MERGE(n0)-[r :T3{k187 : false, k186 : 1795019972, k189 : "4n2TS", k188 : false, **id : 17**, k190 : "v6z5e", **k191 : true**}]->(n1);`
For `a4`, the return should be `true`. (Since r4.k191=true). Therefore, the result should be: `{'a3': 'v6z5e', 'a4': true}`

### Actual behavior
The return value of neo4j is:`{'a3': 'v6z5e', 'a4': 'WEJ6Mifdo'}`
Thank you so much for your help!
![2023-11-30_16-50](https://github.com/neo4j/neo4j/assets/24768784/8f11b5fe-17f7-4d3b-b9fb-8605ec2ac9f7)

**Comment by pontusmelke** [comment]
Hi @Aurora-yzj 

Thanks for reporting this, we will have a look and get back to you!

Regards,
Pontus

**Comment by JonasHolgersson** [comment]
Hi @Aurora-yzj 
This will be fixed in the 5.17 release

Regards,
Jonas

**Comment by Aurora-yzj** [comment]
> Hi @Aurora-yzj This will be fixed in the 5.17 release
> 
> Regards, Jonas

Thanks! May I ask in which commit you fixed this bug? I also have found some other potential correctness issues, and it would be great if I could build a bug-fixed version and verify if all these bugs are introduced by this same root cause.

**Comment by JonasHolgersson** [comment]
> > Hi @Aurora-yzj This will be fixed in the 5.17 release
> > Regards, Jonas
> 
> Thanks! May I ask in which commit you fixed this bug? I also have found some other potential correctness issues, and it would be great if I could build a bug-fixed version and verify if all these bugs are introduced by this same root cause.

The fix will not be in this repository until 5.17 has been released which should be in mid- to late February. There will be a branch called 5.17.
